src/libmeasurement_kit/libevent/dns_impl.hpp

```diff
@@ -53,37 +54,66 @@ class QueryContext : public NonMovable, public NonCopyable {
     }
 };
 
+struct evdns_base_deleter {
+    void operator()(evdns_base* p) {
+        constexpr int fail_requests = 1;
+        evdns_base_free(p, fail_requests);
+    }
+};
+
+struct evaddrinfo_deleter {
+    void operator()(evutil_addrinfo* p) {
+        evutil_freeaddrinfo(p);
+    }
+};
+
+using evdns_base_uptr = std::unique_ptr<evdns_base, evdns_base_deleter>;
+using evaddrinfo_uptr = std::unique_ptr<evutil_addrinfo, evaddrinfo_deleter>;
+
 template <MK_MOCK(evdns_base_new), MK_MOCK(evdns_base_nameserver_ip_add),
         MK_MOCK(evdns_base_free), MK_MOCK(evdns_base_set_option)>
 static inline evdns_base *create_evdns_base(
         Settings settings, Var<Reactor> reactor = Reactor::global()) {
 
-    evdns_base *base;
     event_base *evb = reactor->get_event_base();
+    const int initialize_nameservers = settings.count("dns/nameserver") ? 0 : 1;
+    evdns_base_uptr base(evdns_base_new(evb, initialize_nameservers));
+    if (!base) {
+        throw std::bad_alloc();
+    }
 
-    if (settings.find("dns/nameserver") != settings.end()) {
-        if ((base = evdns_base_new(evb, 0)) == nullptr) {
-            throw std::bad_alloc();
+    if (settings.count("dns/nameserver")) {
+        // libevent can't handle link-local IPv6 nameserver in
+        // evdns_base_nameserver_ip_add, and there is no way to parse alike
+        // addresses in platform-independent way, so that's why getaddrinfo().
+        // fe80::227:22ff:fe45:3a92%wlan0 -- Android
+        // fe80::227:22ff:fe45:3a92%wlp1s0 -- Ubuntu Linux
+        evutil_addrinfo hints = { };
```

Comment:
I've seen that using `struct addrinfo ainfo = {}` causes a g++ warning where the compiler complains that fields are not initialized (specifically with `-Wextra` also `-Wmissing-field-initializers` is enabled by g++ and the warning says that the initializer is missing). Yet the assembly contains a `req stosq` instruction that IIRC should clear the buffer:

```C++
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>

int main() {
    addrinfo hints{};
}
```

and (snipped assembly that should mean "for 6 repetitions, store 0 into memory pointed to by rdi incrementing rdi by 8 each time"):

```amd64
        leaq    -48(%rbp), %rdx
        movl    $0, %eax
        movl    $6, %ecx
        movq    %rdx, %rdi
        rep stosq
```

This seems similar to the assembly generated by memset().

Regarding `explicit_memset()`, the issue arises when you want to clear a private key once you have finished using it, because there is an optimization calling "dead store elimination" where the compiler removes write operations to areas of memory that are no longer used after such store.

This is a concern for private keys because they are cleared after use, but not a concern when you are initializing memory for using it, because that does not trigger the optimization.

Interestingly, [explicit memset is implemented by calling a "weak" function after zeroing the memory](https://github.com/libressl-portable/openbsd/blob/master/src/lib/libc/string/explicit_bzero.c), to fool the optimizer and avoid triggering dead store elimination (and there is also [a discussion on when this approach could fail](https://github.com/libressl-portable/openbsd/issues/5)).

Bottom line: let's just use `memset()`.